core/iwasm/aot/aot_loader.c

```diff
@@ -2532,6 +2535,62 @@ load_init_data_section(const uint8 *buf, const uint8 *buf_end,
     return false;
 }
 
+#if defined(BUILD_TARGET_AARCH64)
```

Comment:
Merging data and text may also benefit other targets, e.g., it may reduce the tlb cache miss and improve the performance since two data may use a same page. How about we make it as a configurable feature, and disable it by default? For example:

add WAMR_BUILD_AOT_MERGED_MMAP in build-scripts/config_common.make,
add WASM_ENABLE_AOT_MERGED_MMAP macro, and set it to 0 by default in core/config.h,
and use WASM_ENABLE_AOT_MERGED_MMAP to control the code in this file?

And here when `merge_data_and_text` fails, just output an warning (e.g., LOG_WARNING), but not return false.

@yamt, how do you think?

Reply:
it makes sense.
i'm not sure if it's worth a config though.

i guess we can unconditionally merge many of data sections at least.
llvm tends to produce multiple rodata sections for objects with different characteristics.
the following is taken from an 1MB x86-64 aot file i happened to have.
as you see, these sections are usually not so large.
using os_mmap for each of these small sections is a bit wasteful.
ideally we can merge them during aot-compilation.
```
data_section_count 5
data_section_name .rodata.cst16
data_section_size 736
data section [0] offset 44924 size 736
data_section_name .rodata
data_section_size 4988
data section [1] offset 45676 size 4988
data_section_name .rodata.cst8
data_section_size 680
data section [2] offset 50684 size 680
data_section_name .rodata.cst4
data_section_size 244
data section [3] offset 51384 size 244
data_section_name .rodata.cst32
data_section_size 448
data section [4] offset 51648 size 448
```

if we always merge (or at least place together) some of sections, i guess we can even resolve some relocations during aot-compilation.

Reply:
Yes, @lum1n0us also suggested to unconditional try to merge data and text, and not to add cmake variable and gcc macro to control the code. It is good to me, my concern is that we should review the code carefully and do more test, to avoid potential logic error and memory leak.

For merging data sections, I think we don't need to change the behavior in the aot compiler, instead in load_object_data_sections of this file, we can traverse the data sections twice, the first time is to calculate the total size of the data sections and then mmap the merged data sections, then setting each data section in the second time traversing.
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_loader.c#L2404-L2417

@bianchui How about adding below fields in AOTModule:
```C
    uint8 *merged_data_sections;
    uint32 merged_data_sections_size;
    uint8 *merged_data_text_sections;
    uint32 merged_data_text_sections_size;
```

And after that, we can mmap the `merged_data_text_sections` for merged `merged_data_sections` and aot text. And normally we had better not do it in two scenarios: (1) the XIP mode, e.g., the aot text is burned in rom device and developer doesn't hope to mmap aot text again in aot loader, (2) the module is loaded by aot_load_from_sections, in which the aot text is pre-mmapped. Note that aot_load_from_sections calls load_from_sections with `is_load_from_file_buf = false`.

Reply:
> For merging data sections, I think we don't need to change the behavior in the aot compiler,

we need to change the compiler to merge sections for xip.
and it would make the non xip case simpler as well.

Reply:
> my concern is that we should review the code carefully and do more test, to avoid potential logic error and memory leak.

i suppose having conditionals actually increases review and testing effort.

Reply:
> we need to change the compiler to merge sections for xip. and it would make the non xip case simpler as well.

But this may require changing the AOT format?

> i suppose having conditionals actually increases review and testing effort.

Agree, I just thought that adding a flag and disabling it by default won't impact the current solution. But it is good not to add the flag if we are confident that the new solution will be well-implemented.

Reply:
should i merge data first into merged_data_sections
and then merge all into merged_data_text_sections add free merged_data_sections ?

> 
> @bianchui How about adding below fields in AOTModule:
> 
> ```c
>     uint8 *merged_data_sections;
>     uint32 merged_data_sections_size;
>     uint8 *merged_data_text_sections;
>     uint32 merged_data_text_sections_size;
> ```
>

Reply:
Yes, that would be better, the data sections are merged even when the later merging of them with aot text fails.

BTW, could you merge data and text unconditionally per suggested by @yamt?
And since the patch may impact all the targets and platforms, I think we had better do more tests, how about I create a development branch e.g. `dev/merge_aot_data_text` and you change this PR to submit to that branch instead? We may test more and merge it into main branch when it is stable.

Reply:
> And after that, we can mmap the `merged_data_text_sections` for merged `merged_data_sections` and aot text. And normally we had better not do it in two scenarios: (1) the XIP mode, e.g., the aot text is burned in rom device and developer doesn't hope to mmap aot text again in aot loader, (2) the module is loaded by aot_load_from_sections, in which the aot text is pre-mmapped. Note that aot_load_from_sections calls load_from_sections with `is_load_from_file_buf = false`.

but load_object_data_sections_info() do the mmap in load_object_data_sections()
but that function do not consider is_load_from_file_buf

Reply:
> Yes, that would be better, the data sections are merged even when the later merging of them with aot text fails.
> 
> BTW, could you merge data and text unconditionally per suggested by @yamt? And since the patch may impact all the targets and platforms, I think we had better do more tests, how about I create a development branch e.g. `dev/merge_aot_data_text` and you change this PR to submit to that branch instead? We may test more and merge it into main branch when it is stable.

ok, i can do it. when the branch created, i will change PR merge target

Reply:
> > And after that, we can mmap the `merged_data_text_sections` for merged `merged_data_sections` and aot text. And normally we had better not do it in two scenarios: (1) the XIP mode, e.g., the aot text is burned in rom device and developer doesn't hope to mmap aot text again in aot loader, (2) the module is loaded by aot_load_from_sections, in which the aot text is pre-mmapped. Note that aot_load_from_sections calls load_from_sections with `is_load_from_file_buf = false`.
> 
> i'm trying to match the `two scenarios` but mmap is being done in load_object_data_sections() and do not consider is_load_from_file_buf

Yes, in load_object_data_sections(), we can just traverse two times, one is to get total size, the other is to mmap and set each data section, no need to consider is_load_from_file_buf. For the two scenarios, I meant we should add checks before calling merge_data_and_text: (1) module is XIP, don't call merge_data_and_text if `module->is_indirect_mode` is true; (2) module is not load from file, since is_load_from_file_buf is provided in load_from_sections, you can extend load_text_section to take that argument and then: load_from_sections(.., is_load_from_file_buf) -> load_text_section(.., is_load_from_file_buf) -> don't call merge_data_and_text if is_load_from_file_buf is false.

Reply:
ok, clear.

Reply:
> > Yes, that would be better, the data sections are merged even when the later merging of them with aot text fails.
> > BTW, could you merge data and text unconditionally per suggested by @yamt? And since the patch may impact all the targets and platforms, I think we had better do more tests, how about I create a development branch e.g. `dev/merge_aot_data_text` and you change this PR to submit to that branch instead? We may test more and merge it into main branch when it is stable.
> 
> ok, i can do it. when the branch created, i will change PR merge target

The branch `dev/merge_aot_data_text` was created: https://github.com/bytecodealliance/wasm-micro-runtime/tree/dev/merge_aot_data_text.

Reply:
> Merging data and text may also benefit other targets, e.g., it may reduce the tlb cache miss and improve the performance since two data may use a same page. 

merged sections align to page_size and do not share page because .text and .data have protection difference for now.